### PR TITLE
Loaders: Consolidate text decoding.

### PR DESCRIPTION
**Pull request by donmccurdy**
Creates `THREE.Loader.decodeText( array )` utility function. See #12642.

### CODE REVIEW

**Comment by Mugen87** [comment]
I really love this change since we save a lot of redundant code! 👍

**Review by Mugen87 — APPROVED** [review]

File: examples/js/loaders/deprecated/LegacyGLTFLoader.js [hunk 1]
```diff
@@ -664,29 +664,6 @@ THREE.LegacyGLTFLoader = ( function () {
...
-
-			s += String.fromCharCode( array[ i ] );
-
-		}
-
-		return s;
-
-	}
-
 	// Three.js seems too dependent on attribute names so globally
```
Comment (takahirox) [inline examples/js/loaders/deprecated/LegacyGLTFLoader.js]: Do you folks think we should keep updating `LegacyGLTFLoader`?

Comment (Mugen87) [reply examples/js/loaders/deprecated/LegacyGLTFLoader.js]: see https://github.com/mrdoob/three.js/issues/12367#issuecomment-338784517

I would say yes, at least for the moment.

**Comment by Mugen87** [comment]
Looks like there is a conflict now because of #12677

**Comment by donmccurdy** [comment]
> Looks like there is a conflict now because of #12677

Rebased 👍

**Comment by mrdoob** [comment]
So... In the not so distant future, I want to turn that `THREE.Loader` into a abstract class that all loaders extend from and implement the usual methods (and also save on redundant code 😅).

So, I think it would be better to create something like `THREE.LoaderUtils` for this.

**Comment by Mugen87** [comment]
> In the not so distant future, I want to turn that THREE.Loader into a abstract class that all loaders extend from and implement the usual methods.

That makes absolutely sense! But in this case, why not adding `.decodeText()` as an instance method of `Loader`? Similar to `. extractUrlBase()`. This can be change now and the call would look like this:

```javascript
var text = THREE.Loader.prototype.decodeText( array );
```

When we refactor the loaders as @mrdoob suggested, it would turn into:
```javascript
var text = this.decodeText( array );
```

**Comment by donmccurdy** [comment]
Sounds good to me — I'd be happy with starting `src/loaders/LoaderUtils` or making `decodeText()` an instance method, either way.

**Comment by mrdoob** [comment]
> That makes absolutely sense! But in this case, why not adding `.decodeText()` as an instance method of `Loader`? Similar to `.extractUrlBase()`. This can be changed now and the call would look like this:

Because both `decodeText()` and `extractUrlBase()` are more like utility methods.

I think this is how `THREE.Loader` should look like:

```js
import { DefaultLoadingManager } from './LoadingManager.js';

function Loader( manager ) {

	this.manager = ( manager !== undefined ) ? manager : DefaultLoadingManager;

}

Object.assign( Loader.prototype, {

	crossOrigin: 'Anonymous',

	load: function () {},

	setCrossOrigin: function ( value ) {

		this.crossOrigin = value;
		return this;

	},

	setPath: function ( value ) {

		this.path = value;
		return this;

	}

} );

export { Loader };
```

**Comment by Mugen87** [comment]
Okay, you convinced me :blush:

BTW: No parsing method in ‘Loader‘, right?

**Comment by mrdoob** [comment]
> BTW: No `.parse()` method in `Loader`, right?

That I don't know. I left it out because I wasn't sure... I guess we should have it for consistency?

**Comment by Mugen87** [comment]
I vote for adding it. Not all loaders actually need the `.parse()` method (like `AudioLoader` or `SVGLoader`) but this is also true for `.setCrossOrigin()`. So i think it's good for consistency reasons to define `.parse()` in `Loader`.

**Comment by looeee** [comment]
Should loaders also have a consistent return type as well? Again, this won't apply to all loaders (`AudioLoader`, `TextureLoader` etc), but most loaders return a subset of 

* models
* scenes
* geometries
* cameras
* lights
* materials
* textures
* animations

And different loaders currently deal with this in a variety of ways. E.g. `ColladaLoader` and `GLTFLoader` return an object with these all easily accessible, while `FBXLoader` returns a single Group.

**Comment by mrdoob** [comment]
> Should loaders also have a consistent return type as well?

I assume you are talking about the `parse()` method?

**Comment by looeee** [comment]
> I assume you are talking about the parse() method? 

Yes, exactly.

**Comment by mrdoob** [comment]
Yep! 👍

**Comment by donmccurdy** [comment]
Ok, `THREE.LoaderUtils` added. ☑️

File: src/loaders/Loader.js [hunk 2]
```diff
@@ -69,18 +69,6 @@ Object.assign( Loader.prototype, {
...
-		var parts = url.split( '/' );
-
-		if ( parts.length === 1 ) return './';
-
-		parts.pop();
-
-		return parts.join( '/' ) + '/';
-
-	},
-
```
Comment (Mugen87) [inline src/loaders/Loader.js]: For better backwards compatibility, we should add an entry in `Three.Legacy.js`. Something like:

```javascript
Object.assign( Loader.prototype, {

	extractUrlBase: function ( url ) {

		console.warn( 'THREE.Loader: .extractUrlBase() has been deprecated. Use THREE.LoaderUtils.extractUrlBase() instead.' );
		return LoaderUtils.extractUrlBase( url );

	}

} );
```

**Review by Mugen87 — CHANGES_REQUESTED** [review]

File: src/loaders/Loader.js (hunk 2 above)
Comment (donmccurdy) [reply src/loaders/Loader.js]: Ok, done!

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**